library/psa_crypto_cipher.c

```diff
@@ -42,39 +42,102 @@ static psa_status_t mbedtls_cipher_validate_values(
     psa_algorithm_t alg,
     psa_key_type_t key_type)
 {
+    (void) key_type; /* May be unused, depending on what algorithms/keys are enabled. */
+
     switch (alg) {
+#if defined(PSA_WANT_ALG_STREAM_CIPHER) || defined(PSA_WANT_ALG_CHACHA20_POLY1305)
+#if defined(PSA_WANT_ALG_STREAM_CIPHER)
         case PSA_ALG_STREAM_CIPHER:
+#endif
+#if defined(PSA_WANT_ALG_CHACHA20_POLY1305)
         case PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_CHACHA20_POLY1305, 0):
-            if (key_type != PSA_KEY_TYPE_CHACHA20) {
-                return PSA_ERROR_NOT_SUPPORTED;
-            }
-            break;
+#endif
+#if defined(PSA_WANT_KEY_TYPE_CHACHA20)
+    if (key_type == PSA_KEY_TYPE_CHACHA20) {
+        break;
+    }
+#endif
+    return PSA_ERROR_NOT_SUPPORTED;
+#endif /* defined(PSA_WANT_ALG_STREAM_CIPHER) || defined(PSA_WANT_ALG_CHACHA20_POLY1305) */
 
+#if defined(PSA_WANT_ALG_CCM) || defined(PSA_WANT_ALG_GCM) || defined(PSA_WANT_ALG_CCM_STAR_NO_TAG)
+#if defined(PSA_WANT_ALG_CCM)
         case PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_CCM, 0):
+#endif
+#if defined(PSA_WANT_ALG_GCM)
         case PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_GCM, 0):
+#endif
+#if defined(PSA_WANT_ALG_CCM_STAR_NO_TAG)
         case PSA_ALG_CCM_STAR_NO_TAG:
-            if ((key_type != PSA_KEY_TYPE_AES) &&
-                (key_type != PSA_KEY_TYPE_ARIA) &&
-                (key_type != PSA_KEY_TYPE_CAMELLIA)) {
-                return PSA_ERROR_NOT_SUPPORTED;
-            }
-            break;
-
+#endif
+#if defined(PSA_WANT_KEY_TYPE_AES) || defined(PSA_WANT_KEY_TYPE_ARIA) || \
+    defined(PSA_WANT_KEY_TYPE_CAMELLIA)
+    if (0
+#if defined(PSA_WANT_KEY_TYPE_AES)
+        || (key_type == PSA_KEY_TYPE_AES)
```

Comment:
Agreed, that is cleaner.

Reply:
I've also gone with `return PSA_SUCCESS` rather than break, and `return PSA_CORRUPTION_DETECTION` if we fall off the end which seemed marginally simpler to follow to me.

Reply:
Yeah, I almost suggested that but then wasn't sure it was important enough. Glad we agree anyway :)